crates/scmi/src/devices/common.rs

```diff
@@ -0,0 +1,531 @@
+    self, DeviceResult, MessageId, MessageValue, MessageValues, ProtocolId, ScmiDevice,
+    ScmiDeviceError, MAX_SIMPLE_STRING_LENGTH, SENSOR_AXIS_DESCRIPTION_GET, SENSOR_CONFIG_GET,
+    SENSOR_CONFIG_SET, SENSOR_CONTINUOUS_UPDATE_NOTIFY, SENSOR_DESCRIPTION_GET, SENSOR_PROTOCOL_ID,
+    SENSOR_READING_GET,
+};
+
+use super::{fake, iio};
+
+// [(NAME, [(PROPERTY, VALUE), ...]), ...]
+pub type DeviceDescription = Vec<(String, DeviceProperties)>;
+type PropertyPairs = Vec<(String, String)>;
+
+#[derive(Debug, Eq, PartialEq, Hash)]
+pub struct DeviceProperties(PropertyPairs);
+
+impl DeviceProperties {
+    pub(crate) fn new(properties: PropertyPairs) -> Self {
+        Self(properties)
+    }
+
+    pub(crate) fn get(&self, name: &str) -> Option<&str> {
+        self.0
+            .iter()
+            .find(|(n, _)| n == name)
+            .map(|(_, v)| v.as_str())
+    }
+
+    fn names(&self) -> HashSet<&str> {
+        self.0.iter().map(|(n, _)| -> &str { n.as_str() }).collect()
+    }
+
+    fn extra<'a>(&'a self, allowed: &[&'a str]) -> HashSet<&str> {
+        let allowed_set: HashSet<&str> = HashSet::from_iter(allowed.iter().copied());
+        self.names().difference(&allowed_set).copied().collect()
+    }
+
+    fn missing<'a>(&'a self, required: &[&'a str]) -> HashSet<&str> {
+        let required_set: HashSet<&str> = HashSet::from_iter(required.iter().copied());
+        required_set.difference(&self.names()).copied().collect()
+    }
+}
+
+pub type MaybeDevice = Result<Box<dyn ScmiDevice>, String>;
+type DeviceConstructor = fn(&DeviceProperties) -> MaybeDevice;
+
+/// Definition of a device kind.
+///
+/// Use [DeviceSpecification::new] to create it.
+pub struct DeviceSpecification {
+    /// Function to call to create the device.
+    ///
+    /// The device properties are those provided on the command line by the
+    /// user.
+    pub(crate) constructor: DeviceConstructor,
+    /// Short description of the device.
+    ///
+    /// Single line, not a complete sentence.
+    short_help: String,
+    /// Long description of the device.
+    ///
+    /// Complete sentences, can span multiple lines.
+    long_help: String,
+    /// Description of the device parameters available to the user.
+    ///
+    /// Each item in the vector corresponds to a single parameter description
+    /// and should start with the parameter name and a followup colon.
+    parameters_help: Vec<String>,
+}
+
+impl DeviceSpecification {
+    /// Creates a new device specification.
+    ///
+    /// See [DeviceSpecification] for the meaning of the arguments.
+    /// The device specification must be used in [available_devices] to
+    /// actually add the device.
+    fn new(
+        constructor: DeviceConstructor,
+        short_help: &str,
+        long_help: &str,
+        parameters_help: Vec<&str>,
+    ) -> Self {
+        Self {
+            constructor,
+            short_help: short_help.to_owned(),
+            long_help: String::from(long_help),
```

Comment:
```suggestion
            long_help: long_help.to_owned(),
```
Similar to the line above

Reply:
Done.

Reply:
Done.